### ISSUE TITLE
"non-void function returning without a value" when returning from a `when`-block

### ISSUE THREAD

**Issue by ckreibich** [zeek/zeek#3839]
I bumped into this while fiddling with `Broker::exists()`, but the following is a shorter example:
```
event zeek_init() {
        when ( local res = lookup_addr(127.0.0.1) ) {
                return;
        }
}
```
When running this:
```
$ zeek ./test.zeek exit_only_after_terminate=T
warning in <params>, line 1: non-void function returning without a value: lambda_<14219442610740974060>
```
This seems wrong, or at least misleading. When printing the function in question at the time of the warning, via `Describe()`, I get:
```
{
if (1 == when-param-12)
        return (to_any_coerceres = lookup_addr(127.0.0.1));

if (2 == when-param-12)
        {
        return ;
        }

return (to_any_coerceT);
}
```
So I suspect that discrepancy in returns is what's triggering the warning. I noticed that `WhenInfo::Build()` has [a comment](https://github.com/zeek/zeek/blob/master/src/Stmt.cc#L1810-L1823) that says this might be addressed elsewhere, but I'm not sure where that might be, or how it works.

@vpax, you're likely closest to this?

**Comment by vpax** [comment]
Interesting, I was just today bumping up against a similar (not the same) problem with `when`s. I'll take this for next month's script optimization maintenance (it's not ZAM-related, but those changes were made in support of script optimization).

**Comment by bbannier** [comment]
Fixed by e960c29acb9d81da69ced871ab5054c3e2152d3b.
